### PR TITLE
[+] Chapter 4

### PR DESCRIPTION
**Pull request by juev**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -307,6 +316,11 @@ data List a
     = Empty
     | Cons a (List a)
 
+instance Functor List where
+    fmap :: (a -> b) -> List a -> List b
+    fmap _ Empty = Empty
+    fmap f (Cons x xs) = Cons (f x) (fmap f xs)
```
Comment (vrom911) [inline src/Chapter4.hs:322]: Yeap! Great recursive implementation ⭐

File: src/Chapter4.hs [hunk 2]
```diff
@@ -472,10 +486,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Trap e <*> _ = Trap e
+    Reward f <*> x = fmap f x
```
Comment (vrom911) [inline src/Chapter4.hs:493]: Awesome reusage of the `Functor` instance 👍🏼

File: src/Chapter4.hs [hunk 3]
```diff
@@ -489,6 +504,7 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+-- TODO
```
Comment (vrom911) [inline src/Chapter4.hs:507]: Hey, do you need some help with this one? 🙂

File: src/Chapter4.hs [hunk 4]
```diff
@@ -629,7 +647,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM ma mb = ma >>= \a -> if a then mb else pure False
```
Comment (vrom911) [inline src/Chapter4.hs:650]: 👏🏼 This is absolutely correct and lazy implementation! Similar to how `&&` works itself! Great job 😊

**Review by vrom911 — APPROVED** [review]
@juev awesome work with such complicated concepts! You did really neat job on operating with Monads 🏆 

I have noticed, that you left list and tree tasks for later, do you need to discuss it first? I am happy to help, just let me know! ❤️
We are really grateful that you came that far in this course with us. Thank you! 🥇

File: src/Chapter4.hs (hunk 3 above)
Comment (juev) [reply src/Chapter4.hs:507]: I think no, thank you! 
I will return to it later.

**Pull request closed without merging**